### PR TITLE
consolidate behavioud of routing in scattermoe kernels

### PR DESCRIPTION
**Pull request by winglian**
# Description

match behaviour of sonicmoe of softmax and sigmoid routing in scattermoe
capture the scattermoe autotuning best kernel metrics so we can prune them better later

## AI Usage Disclaimer
Claude

## Summary by CodeRabbit

* **Refactor**
  * Internal kernel code restructured with centralized helper functions for routing logic and expert computations, reducing code complexity while supporting multiple routing schemes and LoRA configurations.

* **Tests**
  * Comprehensive test coverage added for mixture-of-experts kernels, including multiple routing strategies (softmax and sigmoid variants), shared expert scenarios, and reference implementations for validation against actual performance.

### CODE REVIEW

File: src/axolotl/integrations/kernels/libs/scattermoe_lora/layers.py [hunk 1]
```diff
@@ -220,6 +220,158 @@ def _unwrap_experts_lora(experts_module):
...
+    num_experts = base_gate.num_experts
+    routing_weights, selected_experts = torch.topk(routing_weights, top_k, dim=-1)
+
+    if getattr(base_gate, "norm_topk_prob", True):
+        routing_weights = routing_weights / routing_weights.sum(dim=-1, keepdim=True)
+
+    return routing_weights, selected_experts, top_k, num_experts
+
+
+def _sigmoid_topk_route(
```
Comment (NanoCode012) [inline src/axolotl/integrations/kernels/libs/scattermoe_lora/layers.py:251]: Not sure if we should duplicate code into this repo too? Should we not use a constant utils as done in the kernels folder?

File: tests/integrations/test_scattermoe_lora.py [hunk 2]
```diff
@@ -321,3 +323,389 @@ def test_scatter_moe_gated_mlp_docstring_no_router_logits(self):
...
+
+
+# ============================================================================
+# 7. Routing strategy detection and sigmoid routing
+# ============================================================================
+
+
+def _make_softmax_gate(E=4, H=16, K=2):
+    """Create a mock softmax-style gate (Qwen/OLMoE)."""
+    return SimpleNamespace(
```
Comment (NanoCode012) [inline tests/integrations/test_scattermoe_lora.py:335]: I saw this class before as a suggestion. Does it really properly test what we are using in prod?

File: src/axolotl/integrations/kernels/libs/scattermoe_lora/layers.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/integrations/kernels/libs/scattermoe_lora/layers.py:251]: i think it's fine, this version supports passing in the lora weights for fused ops
